jax/core.py

```diff
@@ -995,48 +998,102 @@ def call_impl(f: lu.WrappedFun, *args, **params):
 call_p.def_impl(call_impl)
 
 
-# ------------------- Jaxpr printed representation -------------------
+# ------------------- Jaxpr checking -------------------
 
-def check_jaxpr(jaxpr: Jaxpr):
-  """Checks well-formedness of a jaxpr.
+def typecheck(aval, x):
+  return typecompat(aval, get_aval(x))
 
-  Specifically it checks that all variabled used are previously defined.
-  """
-  def context():
-    return "\njaxpr:\n{}\n".format(jaxpr)
+def typecompat(aval_ref, aval):
+  """Determine whether `aval` conforms to `aval_ref`"""
+  aval_ref = raise_to_shaped(aval_ref).strip_weak_type()
+  try:
+    return aval_ref == lattice_join(aval_ref, aval).strip_weak_type()
+  except TypeError:
+    return False
 
-  def read_env(env: Set[Var], v: Var):
-    if type(v) is not Literal and v not in env:
-      raise Exception("Variable '{}' not defined".format(v) + context())
+def typematch(aval1, aval2):
+  return (raise_to_shaped(aval1).strip_weak_type() ==
+          raise_to_shaped(aval2).strip_weak_type())
+
+class _JaxprContext(object):
+  __slots__ = ["env", "jaxpr"]
+  def __init__(self, jaxpr: Jaxpr):
+    self.env: Dict[Var, AbstractValue] = {}
+    self.jaxpr = jaxpr
+
+  def _msg_context(self):
+    return "\njaxpr:\n{}\n".format(self.jaxpr)
```

Comment:
Good points. I've removed the need to carry the jaxpr around (only to be appended to error messages everywhere). I kept the class. Because it only represents a type environment now, I renamed it along with the `read`/`write` functions, and added a comment as you suggested.